### PR TITLE
feat(*): Post render hooks for Helm

### PR DESCRIPTION
**Pull request by thomastaylor312**
This adds post-rendering hooks for Helm. See the [proposal issue](https://github.com/helm/helm/issues/7260) for full details

### CODE REVIEW

**Comment by thomastaylor312** [comment]
Here is the associated docs PR for this new feature: https://github.com/helm/helm-www/pull/479

**Review by mattfarina — CHANGES_REQUESTED** [review]
Musing for a moment, this works differently than the downloader getter hook. I'd like to have a consistent pattern to this sort of thing so we can replicate it again. I like this pattern and work here.

In the getter a scheme is registered and the right plugin file to call is looked up from the registration. This also enables plugins to have [platform specific commands](https://helm.sh/docs/topics/plugins/#building-plugins). I'm thinking of Windows here (something the whole CNCF community has not tended to handle well)

Should we make the plugin interaction follow a similar scheme?

File: pkg/postrender/exec.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+func getFullPath(binaryPath string) (string, error) {
+	// Manually check the plugin dir first
+	if !strings.Contains(binaryPath, string(filepath.Separator)) {
+		// First check the plugin dir
+		pluginDir := helmpath.DataPath("plugins") // Default location
+		// If location for plugins is explicitly set, check there
+		if v, ok := os.LookupEnv("HELM_PLUGINS"); ok {
+			pluginDir = v
+		}
+		_, err := os.Stat(filepath.Join(pluginDir, binaryPath))
```
[the remark this answers is not in the document]
Comment (thomastaylor312) [reply pkg/postrender/exec.go]: Let me take a look and see if I could do that. I honestly thought there is always a _single_ plugin directory.

[the remark this answers is not in the document]
Comment (thomastaylor312) [reply pkg/postrender/exec.go]: Let me take a look and see if I could do that. I honestly thought there is always a single plugin directory.

**Comment by thomastaylor312** [comment]
@mattfarina I am not sure I am following what you are are saying here. Is the blocking change on the single comment around handling a list of directories and the comment for the review just a general musing? Or are you saying that if the post-renderer is coming from a plugin directory to have this handle it differently?

File: pkg/postrender/exec.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (thomastaylor312) [reply pkg/postrender/exec.go]: Ok, support (and a test) added

**Comment by technosophos** [comment]
What @mattfarina is suggesting is that you follow the plugin pattern used to extend downloads. Right now, a plugin can declare that it handles a protocol (e.g. `foo://`). When a download request is received, it will check whether any plugins handle that protocol. If so, it will load the plugin, execute it, and pipe the data back into Helm.

I think that the suggestion is to use the plugin pattern for post-render hooks for a few reasons:

- It provides a clear path for multi-platform binaries
- It follows an existing precedent
- It uses the hook system defined in `pkg/plugins`

I believe we chatted about this a long time ago. In my mind, the main way in which this differs from downloaders and other plugins is that essentially what this PR is trying to do is emulate something like `helm-template | $RENDER | helm install` (with where install takes the output of the plugin and directly installs it). `$RENDER`, here, can be any command, and is passed in on the CLI as a flag.

My understanding is that the purpose of this design is to make it super easy for an operator to do one-off shell commands that pipe data through some other existing tool _ad hoc_, not every single time. (Downloaders, in contrast, are automatically loaded and executed whenever necessary -- and the user can be blissfully unaware that this is even happening)

Because I think that the explicit execution of a particular program is the desired "best case", I think the PR is good as it stands. It might be worth considering integrating this into the plugin system in the future, though.

**Review by hickeyma — APPROVED** [review]
LGTM, thanks @thomastaylor312. I think this will be valuable capability for infra folks.

**Comment by mattfarina** [comment]
@technosophos um...

> My understanding is that the purpose of this design is to make it super easy for an operator to do one-off shell commands that pipe data through some other existing tool ad hoc, not every single time.

So, if two people go to upgrade the same application one does it with the script and the other doesn't you can end up with differences.

If the one off case is the expectation than we shouldn't link it into the plugin system at all. If the plugin system is used it should be intentional. If the use case is one-off shall command not regularly used than there is no need for plugins.

I expect this to be abused.

**Comment by technosophos** [comment]
@mattfarina yeah, essentially that is what we are enabling: An opportunity for a chart user to take the output of Helm, pipe it through something like Kustomize, and then install it with Helm. And we do think we want that to be the decision of the individual operator.

I imagine that it could be abused by people doing things like embedding other template languages in their charts. But (a) that can be done today, and (b) I don't think we're in the business of preventing people from making poor life choices.

My argument for including this feature is that it gives a happy path to people who want to use external patch managers like Kustomize. They can still take advantage of the rest of Helm's tooling instead of having to choose whether they want to patch their YAML or whether they want Helm-based installation management.

**Review by bacongobbler — APPROVED** [review]
LGTM! Nicely written, and thank you for the tests.

**Review by mattfarina — APPROVED** [review]
lgtm. I look forward to iterations on this to continue to make this more useful.

**Pull request merged**